cicecore/cicedynB/infrastructure/comm/mpi/ice_global_reductions.F90

```diff
@@ -55,6 +56,12 @@ module ice_global_reductions
                       global_sum_scalar_int
    end interface
 
+   interface global_sums
```

Comment:
I too find this interface a little perplexing.  All the public interfaces in this subroutine are for fields.  This is for a vector of scalars where all scalars are added together across all pes (if I understand correctly).  I would like to suggest that a global_scalar_sum interface would be better and for the local vector sum to be passed into the global_scalar_sum.  I would also call it global_scalar_sum instead of global_sums.  Or if it remains a vector, then we should be able to clearly differentiate an interface where the sum of the vectors produces a vector vs a scalar.  I think we want to differentiate these not just thru different module procedure interfaces, but actually via clearer interface names.  It may be that we never implement the global sum of vectors to vector, but it makes sense to me to put in place a naming convention that makes it clear to do.

Reply:
One other comment.  Is the "global_sums" used for diagnostics or is it needed in the solver?  In either case, I think we need to target the ability to get "bit-for-bit" sums on different decompositions and pe counts using the bfbflag or some other way if possible.  This bit-for-bit capability is important at least for testing.  I haven't looked closely at the branch, but I'd be happy to see if we can leverage the reprosum algorithm to do this.  I think what it also implies is that the per pe values in global_sums cannot already be a local reduction of a field.

Reply:
Ok I'll try to explain what `global_sums` does in a clearer way (I don't mind changing the name):
 In fact it is simply an MPI_ALL_REDUCE. Referencing the interface:

https://github.com/CICE-Consortium/CICE/blob/af8d03e42a4111ee2045f8bf5d7f7a7188637953/cicecore/cicedynB/infrastructure/comm/serial/ice_global_reductions.F90#L713-L714

Here is the computation that's taking place:
```
processes |      "vector"      |  "globalSums"
----------------------------------------------------------------------------------
p1        |  a1, b1, ..., x1   |  a1+a2+...+an, b1+b2+...+bn, ..., x1+x2+...+xn  |
----------------------------------------------------------------------------------
p2        |  a2, b2, ..., x2   |    same as above
----------------------------------------------------------------------------------
...       |    ...             |    same as above
----------------------------------------------------------------------------------
pn        |  an, bn, ..., xn   |    same as above
----------------------------------------------------------------------------------
```


I already use the `compute_dbl_sums` subroutine, so any bfbflag option would already work.
I looked at the code of  `compute_dbl_sums`  and understood that it receives a 2D array, reduces it to a 1D vector by summing along columns, and passes the resulting vector to `MPI_ALL_REDUCE`. So this is exactly what I want, and I simply create a 2D array "`work`" of shape (1,m) in `global_sums_dbl` to accomodate the interface of `compute_dbl_sums`:

https://github.com/CICE-Consortium/CICE/blob/af8d03e42a4111ee2045f8bf5d7f7a7188637953/cicecore/cicedynB/infrastructure/comm/serial/ice_global_reductions.F90#L761-L766

To answer your question, this subroutine is used in the solver if one choose classical Gram-Schmidt as the orthogonalization method.

As for naming, there are already subroutines `global_sum_scalar_{dbl,real,int}` (in the interface `global_sums`) that do a global reduction for scalar values (i.e. a single scalar value on each proc, all summed together across procs to obtain the same scalar values on all procs). So I would suggest that I rename the subroutine to `global_sum_vector_dbl` and the interface to `global_sum_vector` so that everything is clear. Would that work ?

Reply:
Sounds fine.  I think it definitely needs a more descriptive interface name.  Looking at what we have now,

global_sum_dbl    ! sums a field across all pes to form a single scalar
global_sum_scalar_dbl  ! sums a scalar across all pes to form a single scalar

global_sum_vector_dbl might suggest that it is summing a vector across all pes to form a single scalar, but it's not.  What it's really doing is global_sum_scalar_dbl but for multiple scalars at the same time.  In other words, summing a vector across all pes to form a vector.  How about if we call it global_sum_vector_vector_dbl and put it under the generic "global_sum" interface.  Would there be a problem with uniqueness in the interface if we tried to do that?  At the same time, you could implement the global_sum_vector_dbl to acknowledge the difference?

Reply:
My thinking is that since it performs a different kind of action, it makes sense for it to be a separate interface... that's why I'm suggesting not putting it in the generic `global_sum` (there would be no interface conflict if we do, though). Maybe `global_allreduce_sum` (this is closer to `MPI_ALLREDUCE` ) ? This way we would have : 

`global_sum` -> interface to reduce a distributed something (presently, a 2D array or a scalar) to a single scalar
`global_allreduce_sum` -> interface to reduce a distributed something (in this case a 1D vector) to something of the same shape (a 1D vector)

and then the subroutine itself could be `global_allreduce_sum_vector_dbl`. 

What do you think?

Reply:
I could be convinced that there should be a separate generic interface.  One other thing we could do is update the internal module interface names (not global_sum) to make them clearer.  We could come up with a subroutine naming scheme that is more general and extensible.  Maybe 

  global_sum_dbl would become global_sum_field_scalar_dbl
  global_sum_scalar_dbl would become global_sum_scalar_scalar_dbl
  
and the new method could be called

  global_sum_vector_vector_dbl

with room to add

  global_sum_vector_scalar_dbl

in the future if needed.  This should only affect the module and nothing else.  I guess we'd do it in both the serial and mpi versions.  Otherwise, the global_allreduce_sum seems OK for the generic interface name if we keep those methods separate.

Reply:
I went for a separate interface since I feel it's simplest: e30da2b. We can always revisit this in the future.